packages/backend-common/src/scm/git.ts

```diff
@@ -0,0 +1,209 @@
+      onAuth: this.onAuth,
+    });
+  }
+
+  // https://isomorphic-git.org/docs/en/currentBranch
+  async currentBranch({ dir, fullName }: { dir: string; fullName?: boolean }) {
+    const fullname = fullName ?? false;
+    return git.currentBranch({ fs, dir, fullname });
+  }
+
+  // https://isomorphic-git.org/docs/en/fetch
+  async fetch({ dir, remote }: { dir: string; remote?: string }) {
+    const remoteValue = remote ?? 'origin';
+    this.config.logger?.info(
+      `Fetching remote=${remoteValue} for repository {dir=${dir}}`,
+    );
+    return git.fetch({
+      fs,
+      http,
+      dir,
+      remote: remoteValue,
+      onProgress: this.onProgressHandler,
+      headers: {
+        'user-agent': 'git/@isomorphic-git',
+      },
+      onAuth: this.onAuth,
+    });
+  }
+
+  async init({ dir }: { dir: string }) {
+    this.config.logger?.info(`Init git repository {dir=${dir}}`);
+
+    return git.init({
+      fs,
+      dir,
+    });
+  }
+
+  // https://isomorphic-git.org/docs/en/merge
+  async merge({
+    dir,
+    headBranch,
+    baseBranch,
+    author,
+    committer,
+  }: {
+    dir: string;
+    headBranch: string;
+    baseBranch?: string;
+    author: { name: string; email: string };
+    committer: { name: string; email: string };
+  }) {
+    this.config.logger?.info(
+      `Merging branch '${headBranch}' into '${baseBranch}' for repository {dir=${dir}}`,
+    );
+    // If baseBranch is undefined, current branch is used.
+    return git.merge({
+      fs,
+      dir,
+      ours: baseBranch,
+      theirs: headBranch,
+      author,
+      committer,
+    });
+  }
+
+  async push({ dir, remoteName }: { dir: string; remoteName: string }) {
+    this.config.logger?.info(
+      `Pushing directory to remote {dir=${dir},remoteName=${remoteName}}`,
+    );
+    return git.push({
+      fs,
+      dir,
+      http,
+      onProgress: this.onProgressHandler,
+      headers: {
+        'user-agent': 'git/@isomorphic-git',
+      },
+      remote: remoteName,
+      onAuth: this.onAuth,
+    });
+  }
+
+  // https://isomorphic-git.org/docs/en/readCommit
+  async readCommit({ dir, sha }: { dir: string; sha: string }) {
+    return git.readCommit({ fs, dir, oid: sha });
+  }
+
+  // https://isomorphic-git.org/docs/en/resolveRef
+  async resolveRef({ dir, ref }: { dir: string; ref: string }) {
+    return git.resolveRef({ fs, dir, ref });
+  }
+
+  private onAuth = () => ({
+    username: this.config.username,
+    password: this.config.password,
+  });
+
+  private onProgressHandler: ProgressCallback = event => {
+    const total = event.total
+      ? `${Math.round((event.loaded / event.total) * 100)}%`
+      : event.loaded;
+    this.config.logger?.info(`status={${event.phase},total={${total}}}`);
+  };
+}
+
+// TODO(blam): This could potentially become something like for URL
+// and use the integrations config for URLReading instead.
+// But for now, I don't want to do all that in this PR.
+export const fromAuth = ({
```

Comment:
Seeing as we'll prolly need similar logic in the future anyway, should we make the arguments to this `type` and `token` instead of `username` and `password`? That or even something like

```ts
type Options = {
  auth: {
    username: string,
    password: string,
  } | {
    type: 'github' | 'gitlab' | 'azure' | 'bitbucket',
    token: string,
  }
}
```